Updated field-plots standardization with trees and convex hull code.

VictoriaMattsson opened the pull request:
Updated Oren's field-plots code, added trees standardization code, added convex hull gpkg separation for both SSMT Project and DS Project (also renamed files).

VictoriaMattsson committed:
Update 20250312_SSMT_Field_Plots.Rmd

VictoriaMattsson committed:
Update 20250520_SSMT.Rmd

VictoriaMattsson committed:
Update 20250312_DS_Field_Plots_2025.Rmd

VictoriaMattsson committed:
Update 20250520_DS.Rmd

youngdjn commented on contributor_data_to_OFO_schema/20250520_SSMT.Rmd:
```diff
@@ -0,0 +1,320 @@
+      TRUE ~ "0"
+    ),
+    subplots = TRUE,  # unsure if this is accurate
+    subplot_shape = FALSE,
```
Should be blank (`NA`) if there are no subplots. And should never be FALSE, it should describe the shape of the subplots if there are any.

VictoriaMattsson replied:
Ok, subplot_shape should be blank then... but I am questioning if subplots = TRUE. Oren had written that in his code, and he's spent more time with the actual geospatial tree points, so he might have discovered that their plot design matches the ofo subplots definition? But I think it's also possible he saw that there is a PLOT column (we'd call these subplots maybe) along with a SITE column (which we'd call a plot) in the contributed dataset, and didn't investigate further. Because there's no protocol in the metadata beyond a comment/damage codes list, I can't be sure without looking at the geospatial data.

VictoriaMattsson replied:
I see in google docs you said we can go with no subplots or hyperplots, I'll just do that.

youngdjn replied:
Sounds good, yeah given the uncertainty I think that is best. Thanks.

youngdjn commented on contributor_data_to_OFO_schema/20250520_SSMT.Rmd:
```diff
@@ -0,0 +1,320 @@
+    subplots = TRUE,  # unsure if this is accurate
+    subplot_shape = FALSE,
+    subplot_area = "",
+    includes_snags = FALSE,
```
Can you confirm that standing dead trees were ignored? Is this stated in the protocol and/or are there clearly no dead trees in the dataset?

VictoriaMattsson replied:
Good catch, thanks. There was no protocol for this project, but there is an extensive list of damage/comment codes that they used. This included a code for dead trees (DDD), but none were recorded in the dataset. Does this mean includes_snags would be TRUE  since the survey presumably did include snags?

youngdjn replied:
Hm, I'll ask the data contributor about snags. Stay tuned!

youngdjn replied:
I found in the very raw SSMT [data from the contributor](https://drive.google.com/drive/folders/1bXaqeHMGfi-kfVTIVxiLpqhFg0-n92B5?usp=drive_link) that dead trees were surveyed, but not measured, which is about as good as not being surveyed, for our purposes. So you can keep snags as FALSE.

youngdjn commented on contributor_data_to_OFO_schema/20250520_SSMT.Rmd:
```diff
@@ -0,0 +1,320 @@
+    damage_codes_inspected = first(damage_codes_inspected),
+    forest_type = first(forest_type),
+    canopy_cover = first(canopy_cover),
+    min_dbh = min(min_dbh),
```
This is not intended to be the min recorded height. It is intended to be the minimum height necessary to qualify as a tally tree. Thus it should come from the protocol and not the data. And therefore it should be set as one value per plot in line 65 and you could use `first` here to pull one value per plot. Same for the other `min_` and `max_` fields below.

VictoriaMattsson replied:
Oops, remnant from the code before. I don't think it's actually calculating anything since I removed the code that created min_dbh, but I will re-do this whole thing so it's less cluttered.

VictoriaMattsson replied:
And there is no metadata to specify dbh or height cutoffs for trees so I will leave it blank with a comment saying so.

youngdjn replied:
OK and I'm asking the data contributor if there's a min size

youngdjn replied:
I'm holding off on asking the contributor until we know if there are any more questions to ask about the DS study at the same time.

VictoriaMattsson replied:
I think the only question about DS is also whether there is a size cutoff for surveying.

youngdjn replied:
OK I'm sending her these questions now.

youngdjn commented on contributor_data_to_OFO_schema/20250520_SSMT.Rmd:
```diff
@@ -0,0 +1,320 @@
+# Pivot the data to get one row per plot with multiple columns for each field
+SSMT_one_record_per_plot <- SSMT %>%
+  group_by(plot_id) %>%
+  summarise(
```
An alternative to this approach is that you summarize/collapse `SSMT` down to one row per plot first, near the start of the script, before assigning all the values. Because it's not really necessary or elegant to set all these values for every tree in each plot, which is what you're currently doing. Collapsing earlier would avoid the need to apply this summarize step here that takes `first` for every column. In an earlier collapse, you'd only have to keep a few columns, like plot ID (maybe only that), and you could create all the rest in code.

VictoriaMattsson replied:
So true, agree.

youngdjn commented on contributor_data_to_OFO_schema/20250520_SSMT.Rmd:
```diff
@@ -0,0 +1,320 @@
+
+# Notes were recorded in a complicated way. There are two columns (COMM21 and CODE21) which both contain comment codes for damage and other notable features of trees. Sometimes there are multiple codes recorded in one column and sometimes they're separated into the two columns. 
+
+# NOTE: I am ignoring PCDOE and PCOMM, which is the same thing but for prior to 2021. It would be complicated to include these because there is redundancy and it is unclear when these were recorded, thus whether the note is still relevant to the trees as of 2021.
```
That sounds like the right call!

youngdjn commented on contributor_data_to_OFO_schema/20250520_SSMT.Rmd:
```diff
@@ -0,0 +1,320 @@
+  stringsAsFactors = FALSE
+)
+
+annotate_codes_simple <- function(text, dict) {
```
Nice work putting chatgpt to use for this task!

youngdjn commented on contributor_data_to_OFO_schema/20250520_SSMT.Rmd:
```diff
@@ -0,0 +1,320 @@
+
+# convert units
+SSMT <- SSMT %>%
+  mutate(DBH21_cm = DBH21 / 10, # convert DBH from mm to cm
+         CWA21_m = CWA21 / 100, # CHECK! assuming CW_21 is in cm, converting to m
+         CWB21_m = CWB21 / 100) # CHECK! assuming CW_21 is in cm, converting to m
```
Do you know that DBH is in mm (i.e. metric)? If so then I feel comfortable assuming crown width is in cm since the resulting values are in the right magnitude. The main thing to rule out is that they are not in inches, which they should not be if we know DBH is mm.

VictoriaMattsson replied:
I only know because Oren left a very clear/intentional note at the top of his code:

# HT21 Field: Original measurement values recorded in (cm) converted to (m) 
# PHT Field: Original measurement values recorded in (cm); no conversion 
# PDBH Field: Original measurement values recorded in (mm); no conversion
# DBH21 Field: Original measurement values recorded in (mm); no conversion

So I feel like he must've gotten that info from somewhere reliable-- would you happen to know if he reached out to someone? Again, there's no protocol included, so I can't verify. I looked at a few trees to make sure the values make sense, and the height to dbh to crown width ratios seem reasonable when in the units we're assuming here.

youngdjn replied:
Sweet, that's pretty clear! We can trust it. He was in touch with the data contributor and must have asked. Also maybe that raw data folder I linked from google drive above stated it.

youngdjn commented on contributor_data_to_OFO_schema/20250520_SSMT.Rmd:
```diff
@@ -0,0 +1,320 @@
+    percent_prefire_crown_green = "",
+    percent_postfire_crown_green = "",
+    live_crown_class = "",
+    crown_width_1 = CWA21_m, # assuming these two are perpendicular to each other, metadata does not say
+    crown_width_2 = CWB21_m, # ^
```
Good comment. This is sufficient for documentation of uncertainty here.

youngdjn commented on contributor_data_to_OFO_schema/20250520_DS.Rmd:
```diff
@@ -0,0 +1,274 @@
+      TRUE ~ "0"
+    ),
+    subplots = FALSE,
+    subplot_shape = FALSE,
```
Leave `subplot_shape` undefined or `NA` when there are no subplots

youngdjn commented on contributor_data_to_OFO_schema/20250520_DS.Rmd:
```diff
@@ -0,0 +1,274 @@
+    subplots = FALSE,
+    subplot_shape = FALSE,
+    subplot_area = "",
+    includes_snags = FALSE,
```
Does the protocol explicitly say snags are excluded? Or is it possible some of Oren's pre-processing accidentally removed them?

VictoriaMattsson replied:
That's an interesting point, I was going based off this description of the dataset in the metadata: "Contains tree measurements and "local" stem-mapped coordinates for live trees in 2013." which specifies live trees, so my assumption is based on that. There's also no code or designation for dead trees. 
After some investigation just now, I found in the matadata a link to some more publicly available, which seems to include this dataset I've been using plus more years of data, also a MortalityYear column. I can try joining that column, but from a quick look at this expanded dataset, it seems possible that some (but not all?) dead trees were removed from the dataset I've been working with. I'll look into it more.

youngdjn replied:
Sweet, let me know what you find! Please be alert to the possibility that trees with a MortalityYear were already filtered out in Oren's processing. I just found more very raw data [here](https://drive.google.com/drive/folders/10mnLYFkJB44x4l4n4aV1z_DX2VAzSsBR?usp=drive_link) that might provide more clues, or are maybe what you found yourself.

VictoriaMattsson replied:
I haven't seen those until now, I found my new dataset through the link below. It looks like the data in the folder you shared is already almost as processed as what I've been using (Oren removed 24 trees, those outside of plot boundaries or on the ground). So now I'm thinking that maybe they were filtered out before Oren began working with the data-- is that possible?

link to the new data I found:
in box, this doc lists some public data and publications related to this project: https://ucdavis.app.box.com/file/1801604575095
and then if you download the files from that page (near the bottom), there's a dataset that has these trees with measurements for other years besides 2013 (so that there are 24,000 rows instead of 3,0000 because the data is pivoted in an interesting way). At quick glance there are 500ish more trees in this dataset.

If I find out that dead trees were removed by somebody (have a feeling this is possible, need to look more), should I try to get them back?

youngdjn replied:
Wow good find, I wonder why we weren't just using that published data all along, especially since it has a more recent survey. If you still have the time and interest, then yes, ideally you would pull in the dead trees if possible (though this might be hard because you'd also have to get them into geospatial coordinates) and also pull in the more recent (2015) measurements.

A caveat regarding the dead trees: if we're going to include them, we need to know that they were standing, and that the height associated with them was measured in the most recent survey (e.g., if a snag's top broke so the tree became shorter, the height measurement should reflect that). If that's not possible, then we should no pull in snags, and set includes_snags to FALSE.

Even if there is not a way to pull in snags, it would still be good to pull in the more recent 2015 measurements (and maybe a few more trees died by 2015 that could then be excluded).

VictoriaMattsson replied:
Great! I just used the 2015 height, dbh, height to live crown, damage, and lean measurements instead of our contributed dataset measurements taken in 2013. I'm keeping Oren's adjusted coords from the contributed dataset (also checked to make sure things like tree id and species and treatment were consistent across both, they are). 

It looks like 671 dead trees were intentionally removed from the dataset (the only not present in both our received data and the 2015 data I found all have a mortality year! very clear), and only a few of those are on the ground (there's a column with a code for this, will remove downed trees), so there are a good amount of dead trees to add to our ofo dataset-- at the end of it all, 522 more trees added because of this dataset, and all measurements 2 years more recently recorded.

The part I'm stuck on is adjusting what seems like plot-referenced coords into geospatially referenced coordinates (tree_lat and tree_lon in the cleaned data, currently left blank for newly added trees)... also what these extra dead trees mean for the geospatial files oren has made (don't know much about those)... It's not something I could readily know how to do, but am happy to work with you to figure it out!

youngdjn replied:
It's great you pulled in the 2015 measurements for the live trees! The geospatial coords will be a lot harder, and not possible in the time available. It involves plotting the trees out in their internal coordinate system in GIS, then dragging/rotating them around the map until they align with the drone data. It is unavoidably manual, not coded. This could be a good task for a possible extension to your position. For now, please remove the dead trees again (sorry!) but leave clear comments in the code to guide us in adding them back later.

youngdjn commented on contributor_data_to_OFO_schema/20250520_DS.Rmd:
```diff
@@ -0,0 +1,274 @@
+    max_dbh_of_primary_trees = "",
+    min_ht = "",
+    min_dbh_ohvis = "",
+    min_ht_ohvis = "",
```
See my comments about these fields on the SSMT script. There should be protocol-specified tally tree cutoff(s)

VictoriaMattsson replied:
I don't see it in the provided metadata, since it seems like just an overview of this dataset and what the codes mean, rather than an actual protocol. I do know that they only measured height for their treatment trees. They did also measure ingrowth trees, so dbh can be very small, down to 1.2 cm according to the dataset... which doesn't really make sense if one is measuring at breast height... I don't really know what to make of that, but it seems like they don't have a min height (since it's just for treatment trees) or a min dbh (since it goes down to a centimeter, so maybe 1cm is the cutoff? not sure). Thoughts?

youngdjn replied:
I'll ask Connie, the contributor.

youngdjn commented on contributor_data_to_OFO_schema/20250520_DS.Rmd:
```diff
@@ -0,0 +1,274 @@
+    height_above_plot_center = "",
+    species = case_when(
+      SPECIES == "DF" ~ "202", # Douglas-fir
+      SPECIES == "CH" ~ "768", # cherry - NOTE: contributors do not specify species so assuming this is bitter cherry but not confirmed
```
This seems reasonable and having this uncertainty documented here is sufficient.

youngdjn commented on contributor_data_to_OFO_schema/20250520_DS.Rmd:
```diff
@@ -0,0 +1,274 @@
+      SPECIES == "DG" ~ "492", # pacific dogwood
+      SPECIES == "MA" ~ "312"), # bigleaf maple
+    growth_form = "tree", # assuming all are trees
+    live_dead = "L", # live trees only
```
How do you know it's live trees only? It seems like there should be at least some dead if some had 90% of bole circumference damaged, seems the intent of the study was to figure out how much damage trees can survive (or not). Is it possible Oren's pre-processing accidentally removed dead trees?

youngdjn commented on contributor_data_to_OFO_schema/20250520_DS.Rmd:
```diff
@@ -0,0 +1,274 @@
+    decay_class = "",
+    damage_1 = case_when(
+      HTDAM_13 == "0" ~ "", # none
+      HTDAM_13 == "1" ~ "90004", # fork -> 90004 forked top
```
I'd guess "fork" is more likely 90005 (fork below top). In our damage codes defs I'll add that 90005 can also simply mean "fork".

youngdjn commented on contributor_data_to_OFO_schema/20250520_DS.Rmd:
```diff
@@ -0,0 +1,274 @@
+    height = HT_13,
+    dbh = DBH_13,
+    height_to_crown = HLC_13,
+    contributor_tree_id = TREE) %>%  # note that contributor did not use unique id's, TREE nums reset each plot, also ingrowth trees are labeled with TREE num of nearest existing tree (shown in RECRUIT col of original data)
```
Thanks for adding this useful comment!

youngdjn reviewed (changes requested):
This is all very very good work -- great attention to detail. Thank you! I have some pretty straightforward inline comments. The biggest ones involve (1) confirming that exclusion of snags was real and intentional, and not just for example an accident in the upstream data prep of the input files you are working with and (2) getting the protocol-specified tally tree threshold(s) rather than computing the min tree size from the data. Remember that min_ht and min_dbh (etc) are not intended to summarize the observations, but to indicate the cutoffs in the survey protocol.

youngdjn commented on contributor_data_to_OFO_schema/20250520_DS.Rmd:
```diff
@@ -0,0 +1,274 @@
+    crown_width_2 = "",
+    crown_width_allometric = "",
+    decay_class = "",
+    damage_1 = case_when(
```
The damage codes I see in the data are all only 90001 (lots), 90004 (one tree), and no others. It seems we should at least see 90006 at least once. Can you please check that it is correct that nearly all the damages are 90001 and there is no 90006? The way it's scripted here, it's possible that 90006 only occurs on a tree that also has another damage (coded in `HTDAM`), in which case the other damage code (whatever comes first in the `case_when`) would take precedence and override the 90006.

VictoriaMattsson replied:
I'm not sure exactly what you're asking, but maybe this clarifies some things: there are 55 trees with leandown = 3, which I'd reclassify as 90006, it's just commented out right now so that didn't make it to the output csv, which is what I'm assuming you searched for 90006 in.

youngdjn replied:
Ah! I think I missed or forgot that it was commented out. Uncommenting it should be good! However we still would have the issue with potentially multiple damage codes per tree: in this case, only the first code would be populated, as currently written.

VictoriaMattsson replied:
Ok good point, I will make sure that doesn't happen.

VictoriaMattsson replied:
Good point, will make sure that doesn't happen, thanks!

VictoriaMattsson committed:
Responses to comments on DS and SSMT projects.

Derek gave feedback on my first round of DS and SSMT standardization, these updates reflect those comments. The major one which leaves future work is that DS gained roughly 500 dead trees which were not originally in contributed data-- these have not been georeferenced and are missing coords right now.

youngdjn commented on contributor_data_to_OFO_schema/20250520_DS.Rmd:
```diff
@@ -174,26 +167,72 @@ test <- DS %>%
+
+# designating all treated trees as damage = mechanical 
+mechanical_codes <- c("B100-1S", "B20-1S", "B40-1F", "B40-1S", "B40-2S", "B60-1S", "B80-1S", "B80-2S", "B90-1S", "R25", "R50", "R25-B40-1S", "R50-B40-1S", "PRUNE", "SOIL") 
+# is soil removal and replacement "mechanical"? I'm calling it that because still human activity damage
```
Good question; I don't think we should include this. We only care about damage to the aboveground part of the tree.

youngdjn commented on contributor_data_to_OFO_schema/20250520_DS.Rmd:
```diff
@@ -174,26 +167,72 @@ test <- DS %>%
+
+combined <- combined %>%
+  mutate(
+    dmg_ht = case_when(
+      HeightDamage.code == 1 ~ "90005", # fork -> 90005 fork below top ## WHY SO MANY OF THESE TREES?
+      HeightDamage.code == 2  ~ "90002", # dead top -> 90002 dead top
+      HeightDamage.code == 3 ~ "90001", # broken top -> 90001 broken top
+      TRUE ~ NA_character_),
```
Is it not possible to have more than one height damage (at least in this dataset)? This approach would only pull one damage, giving priority to  90005, then 90002, then 90001. Note that if you change this so `dmg_ht` contains a vector/list of damages (instead of a single value like "90005"), then those multiple values will be passed through the `coalesce` below, resulting in a final damage column with multiple values.

One idea is: across all possible damages (height (possibly multiple), trt, and lean), combine them all into a vector (or a comma-separate string like "90005,90002,90006"), then split this string out into the 3 columns. In the case of the comma-separated string, I think `dplyr::separate_wider_delim` would do this.

If this would take more time than you have, stick with your current approach but keep track of this issue if you can so that we can return to it if you are able to return to this position.

VictoriaMattsson replied:
The HeightDamage.code column is only ever populated with one value, so I don't think this should be an issue!

youngdjn replied:
OK great!

youngdjn commented on contributor_data_to_OFO_schema/20250520_DS.Rmd:
```diff
@@ -174,26 +167,72 @@ test <- DS %>%
+      ifelse(!is.na(dmg_trt) & damage_1 != dmg_trt, dmg_trt, NA_character_),
+      ifelse(!is.na(dmg_lean) & damage_1 != dmg_lean, dmg_lean, NA_character_)
+    ),
+    damage_3 = coalesce(
+      ifelse(!is.na(dmg_ht) & damage_1 != dmg_ht & damage_2 != dmg_ht, dmg_ht, NA_character_),
+      ifelse(!is.na(dmg_trt) & damage_1 != dmg_trt & damage_2 != dmg_trt, dmg_trt, NA_character_),
+      ifelse(!is.na(dmg_lean) & damage_1 != dmg_lean & damage_2 != dmg_lean, dmg_lean, NA_character_)
```
Brilliant!!

youngdjn commented on contributor_data_to_OFO_schema/20250520_DS.Rmd:
```diff
@@ -207,35 +246,28 @@ DStrees <- DS %>% mutate(
+    height = HT.m,
+    dbh = DBH.cm,
+    height_to_crown = HLC.m,
+    contributor_tree_id = TreeNum) %>%  # note that contributor did not use unique id's, TreeNum resets each plot, also ingrowth trees are labeled with TreeNum of nearest existing tree (shown in Recruit col of original data)
```
Good comment. Can you please add if true: within a plot, tree ID is unique.

VictoriaMattsson replied:
Hm well tree ID is not unique within a plot because ingrowth trees are labeled with TreeNum of the nearest previously tagged tree. Otherwise, yes TreeNum resets every plot. Hopefully the comment I wrote is clear enough to explain that! The way you phrased it omits the note about ingrowth trees sometimes making for repeat id's within a plot.

youngdjn replied:
Thanks for clarifying what I missed. Looks good.

youngdjn commented on contributor_data_to_OFO_schema/20250520_DS.Rmd:
```diff
@@ -207,35 +246,28 @@ DStrees <- DS %>% mutate(
     percent_postfire_crown_green, live_crown_class, crown_width_1, crown_width_2, crown_width_allometric,
     decay_class, damage_1, damage_2, damage_3, damage_4, damage_5, distance_to_pith, distance_to_face, azimuth, contributor_tree_id, notes, corrected_error))
 
+# we now have 615 dead trees, thanks to the 2015 data (note some trees that were in the 2013 contributed data were dead before 2013, but that's only reflected in the new big ten year dataset, so that's why there are a few more dead trees than the num of trees added from the big dataset, i.e. more info revealed more dead)
```
This is awesome. Even though for now we are not adding the previous dead, it's great to have the new dead be excluded.

youngdjn reviewed (approved):
Great! Please see some comments on new code below, mostly as a record for opportunities to improve on this later if you return to this position, but also a couple quick fixes for now.

youngdjn commented:
@VictoriaMattsson, I think we (you) have now resolved all outstanding issues so this is good to merge, and the resulting processed data can be added to the full data spreadsheet!

VictoriaMattsson committed:
Last few small changes!

Updated min_dbh and min_ht at plot level for both DS and SSMT projects based on our discussions about what to reasonably set thresholds to, since protocols did not have clear strict cutoffs. Our reasoning is described in-line and at the plot level in the google sheet.
